### ISSUE TITLE
Inconsistent handling of $sce trustedUrl with $http

### ISSUE THREAD

**Issue by klhuillier** [angular/angular.js#11328]
```
// Works:
$http.get($sce.trustAsUrl('/')).success(function(d){$log.log(d);});

// Fails:
$http.get($sce.trustAsUrl('/'), {params: {timestamp: 0}}).success(function(d){$log.log(d);});
```

In the first case, when the trusted URL reaches the XmlHttpRequest, its toString method is used to get the wrapped URL. In the second case, the buildUrl function attempts to add parameters, but the given URL is not a string as expected.

I feel that the buildUrl function should reject a non-string argument for url in all cases.

**Comment by petebacondarwin** [comment]
Or perhaps `buildUrl` should understand and cope with "trusted" resources?

**Comment by petebacondarwin** [comment]
@pkozlowski-opensource could you take a look at this as part of the `buildUrl` refactoring?

**Comment by klhuillier** [comment]
I thought about suggesting buildUrl tolerate trusted URLs, but I couldn't think of a reason it should. I also can't think of a reason it should not, other than, "is it really necessary?" I would be agreeable to either option.

The main issue was trusted URLs were working some of the time, but failing with an unhelpful message at other times. According to the current documentation, it should be failing every time because url is specified as type string.

**Comment by chirayuk** [comment]
I have a work in progress PR that will fix this as a side effect of fixing #11352.  I started on it yesterday but didn't have time to finish it.  I won't get to it anytime today or tomorrow.  However, I created a work-in-progress PR to let folks know that some work has started on it.

**Comment by petebacondarwin** [comment]
In December we put in an explicit check for non-string URLs so that effectively closes this issue.
We are still looking at making JSONP more secure with `$sce` which might require us to relax that constraint going forward.
